File: java/350_IntersectionofTwoArraysII.java
class Solution {

    private void swapArr(int[] nums1, int[] nums2) {
        int[] temp = nums1;
        nums1 = nums2;
        nums2 = temp;
    }


    public int[] intersect(int[] nums1, int[] nums2) {
        if (nums1.length < nums2.length) {
            swapArr(nums1, nums2);
        }
        Map<Integer, Integer> mp = new HashMap<>();

        for (int i = 0; i < nums1.length; i++) {
            if (mp.containsKey(nums1[i])) {
                mp.put(nums1[i], mp.get(nums1[i]) + 1);
            }else {
                mp.put(nums1[i], 1);
            }
        }

        List<Integer> list = new ArrayList<>();

        for (int i = 0; i < nums2.length; i++) {
            if (mp.containsKey(nums2[i])) {
                mp.put(nums2[i], mp.get(nums2[i]) - 1);
                if (mp.get(nums2[i]) >= 0) {
                    list.add(nums2[i]);
                }
            }
        }

        int[] res = new int[list.size()];

        for (int i = 0; i < list.size(); i++) {
            res[i] = list.get(i);
        }

        return res;
    }
}

File: c++/350_IntersectionofTwoArraysII.cpp
//sol1
class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        int len1 = (int)nums1.size(), len2 = (int)nums2.size();
        int p1 = 0, p2 = 0;
        vector<int> res;
        while (p1 < len1 && p2 < len2) {
            int num1 = nums1[p1], num2 = nums2[p2];
            if (num1 == num2) {
                res.push_back(num1);
                p1++, p2++;
            } else if (num1 < num2) {
                p1++;
            } else {
                p2++;
            }
        }
        return res;
    }
};
//sol2
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_multiset<int> s;
        vector<int> res;
        for (auto num : nums1) s.insert(num);
        for (auto num : nums2) {
            if (s.count(num)) {
                res.push_back(num);
                s.erase(s.find(num));
            }
        }
        return res;
    }
};